File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/plugin/js/JSPromotion.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in the root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.plugin.js

import com.getcapacitor.JSObject
import com.microblink.core.Promotion
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents a promotional item as a response in the TIKI SDK.
 *
 * @property promotion The [Promotion] object to convert into an JSPromotion.
 */
class JSPromotion(promotion: Promotion) {
    private val id: Long
    private val slug: String?
    private val reward: Double?
    private val rewardCurrency: String?
    private val errorCode: Int
    private val errorMessage: String?
    private val relatedProductIndexes: List<Int>
    private val qualifications: List<List<Int>>

    init {
        id = promotion.id()
        slug = promotion.slug()
        reward = promotion.reward()?.toDouble()
        rewardCurrency = promotion.rewardCurrency()
        errorCode = promotion.errorCode()
        errorMessage = promotion.errorMessage()
        relatedProductIndexes = promotion.relatedProductIndexes() ?: emptyList()
        qualifications = promotion.qualifications() ?: emptyList()
    }

    /**
     * Converts the JSPromotion object into a JSON representation.
     *
     * @return A [JSONObject] containing the JSON representation of the JSPromotion.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("id", id)
            .put("slug", slug)
            .put("reward", reward)
            .put("rewardCurrency", rewardCurrency)
            .put("errorCode", errorCode)
            .put("errorMessage", errorMessage)
            .put("relatedProductIndexes", JSONArray(relatedProductIndexes))
            .put("qualifications", JSONArray(qualifications.map { q -> JSONArray(q) }))

    companion object {
        /**
         * Converts a [Promotion] object into an [JSPromotion] object.
         *
         * @param promotion The [Promotion] object to convert.
         * @return An [JSPromotion] object or null if the input is null.
         */
        fun opt(promotion: Promotion?): JSPromotion? =
            if (promotion != null) JSPromotion(promotion) else null
    }
}

File: ios/Plugin/Plugin/Js/JSPromotion.swift
/*
 * RspPromotion Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor

/**
 Represents a response containing promotion information.

 This struct is used to convey details about promotions, including the promotion's slug, reward amount, error code, error message, related product indexes, and qualifications.
 */
struct JSPromotion {
    /// The slug of the promotion, if available.
    private let slug: String?
    /// The reward amount associated with the promotion, if available.
    private let reward: Float?
    /// The currency of the reward amount, if available.
    private let rewardCurrency: String?
    /// The error code associated with the promotion.
    private let errorCode: Int
    /// The error message associated with the promotion, if available.
    private let errorMessage: String?
    /// Indexes of related products associated with the promotion.
    private let relatedProductIndexes: [NSNumber]
    /// Qualifications for the promotion.
    private let qualifications: [[NSNumber]]

    /**
     Initializes an `RspPromotion` struct.

     - Parameters:
        - promotion: The `BRPromotion` object containing promotion information.
     */
    init(promotion: BRPromotion) {
        slug = promotion.slug
        reward = promotion.rewardValue
        rewardCurrency = promotion.rewardCurrency
        errorCode = promotion.errorCode
        errorMessage = promotion.errorMessage
        relatedProductIndexes = promotion.relatedProductIndexes
        qualifications = promotion.qualifications
    }
    
    static func opt(promotion: BRPromotion?) -> JSPromotion? {
        return promotion != nil ? JSPromotion(promotion: promotion!) : nil
    }

    /**
     Converts the `RspPromotion` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing promotion information in a format suitable for a Capacitor plugin call result.
     */
    func toJSObject() -> JSObject {
        var ret = JSObject()
        ret["slug"] = slug
        ret["reward"] = reward
        ret["rewardCurrency"] = rewardCurrency
        ret["errorCode"] = errorCode
        ret["errorMessage"] = errorMessage
        ret["relatedProductIndexes"] = JSArray(relatedProductIndexes)
        ret["qualifications"] = JSArray(qualifications.map { q in JSArray(q) })
        return ret
    }
}
